zarr/tests/test_convenience.py

```diff
@@ -85,3 +88,235 @@ def test_lazy_loader():
+        assert a.compressor == baz.compressor
+    assert_array_equal(a[:], baz[:])
+    assert a.attrs['units'] == 'metres'
+
+    # copy array with name
+    dest = new_dest()
+    copy(source['foo/bar/baz'], dest, name='qux')
+    assert 'baz' not in dest
+    a = dest['qux']
+    assert a.dtype == baz.dtype
+    assert a.shape == baz.shape
+    assert a.chunks == baz.chunks
+    if hasattr(a, 'compressor') and hasattr(baz, 'compressor'):
+        assert a.compressor == baz.compressor
+    assert_array_equal(a[:], baz[:])
+    assert a.attrs['units'] == 'metres'
+
+    # copy array, provide creation options
+    dest = new_dest()
+    compressor = Zlib(9)
+    if isinstance(dest, Group):
+        copy(source['foo/bar/baz'], dest, without_attrs=True, compressor=compressor,
+             chunks=True)
+    else:
+        copy(source['foo/bar/baz'], dest, without_attrs=True, compression='gzip',
+             compression_opts=9, chunks=True)
+    a = dest['baz']
+    assert a.dtype == baz.dtype
+    assert a.shape == baz.shape
+    assert a.chunks != baz.chunks  # autochunking was requested
+    if hasattr(a, 'compressor'):
+        assert compressor == a.compressor
+        if hasattr(baz, 'compressor'):
+            assert a.compressor != baz.compressor
+    else:
+        assert a.compression == 'gzip'
+        assert a.compression_opts == 9
+    assert_array_equal(a[:], baz[:])
+    assert 'units' not in a.attrs
+
+    # copy group, default options
+    dest = new_dest()
+    copy(source['foo'], dest)
+    g = dest['foo']  # defaults to use source name
+    assert g.attrs['experiment'] == 'weird science'
+    a = g['bar/baz']
+    assert a.dtype == baz.dtype
+    assert a.shape == baz.shape
+    assert a.chunks == baz.chunks
+    if hasattr(a, 'compressor') and hasattr(baz, 'compressor'):
+        assert a.compressor == baz.compressor
+    assert_array_equal(a[:], baz[:])
+    assert a.attrs['units'] == 'metres'
+
+    # copy group, non-default options
+    dest = new_dest()
+    copy(source['foo'], dest, name='qux', without_attrs=True)
+    assert 'foo' not in dest
+    g = dest['qux']
+    assert 'experiment' not in g.attrs
+    a = g['bar/baz']
+    assert a.dtype == baz.dtype
+    assert a.shape == baz.shape
+    assert a.chunks == baz.chunks
+    if hasattr(a, 'compressor') and hasattr(baz, 'compressor'):
+        assert a.compressor == baz.compressor
+    assert_array_equal(a[:], baz[:])
+    assert 'units' not in a.attrs
+
+    # copy group, shallow
+    dest = new_dest()
+    copy(source, dest, name='eggs', shallow=True)
+    assert 'eggs' in dest
+    eggs = dest['eggs']
+    assert 'spam' in eggs
+    a = eggs['spam']
+    assert a.dtype == spam.dtype
+    assert a.shape == spam.shape
+    assert a.chunks == spam.chunks
+    if hasattr(a, 'compressor') and hasattr(spam, 'compressor'):
+        assert a.compressor == spam.compressor
+    assert_array_equal(a[:], spam[:])
+    assert 'foo' not in eggs
+    assert 'bar' not in eggs
```

Comment:
Would be good to have some error tests here as well. For example, what happens when a Group or Dataset/Array is already present? Do we have any incomplete copies for cases like this or is dest left unchanged?